bpf/kmesh/workload/sockops_tuple.c

```diff
@@ -14,102 +14,156 @@
+ * 6.   |0                  |ip2    |0          |||0            |
+ * 7.   |0                  |ip3    |0          |||0            |
+ *
+ * Why design it that way?
+ * We need a way to mark in the cni whether the current ip is managed by Kmesh.
+ * The cni inserts the ip address into the map when the pod is created and removes the ip
+ * address from the map when the pod is destroyed.
+ * However, according to the cni guide, when deleting the data, only the CONTAINER and IFNAME
+ * (https://github.com/containernetworking/cni.dev/blob/main/content/docs/spec.md#del-remove-container-from-network-or-un-apply-modifications)
+ * must be transferred. The IP address is not transferred in the cni. Therefore, the
+ * containerID and IP address must be bound and stored in the map for subsequent deletion.
+ */
+struct kmesh_manager_key {
+	__u64 containerID;
+	__u32 ipaddress;
+	__u32 index;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, struct kmesh_manager_key);
+	__type(value, __u32);
+	__uint(max_entries, MAP_SIZE_OF_MANAGER);
+	__uint(map_flags, 0);
+} map_of_kmesh_manager SEC(".maps");
 
-static inline void build_auth_key(struct bpf_sock_ops *skops, struct bpf_sock_tuple* key)
+struct {
+	__uint(type, BPF_MAP_TYPE_SOCKHASH);
+	__type(key, struct bpf_sock_tuple);
+	__type(value, __u32);
+	__uint(max_entries, MAP_SIZE_OF_MANAGER);
+	__uint(map_flags, 0);
+} map_of_kmesh_hashmap SEC(".maps");
+
+static inline int is_managed_by_kmesh(__u32 ip)
 {
-    if(key) {
-        key->ipv4.saddr = skops->remote_ip4;
-        key->ipv4.sport = skops->remote_port >> 16;
-        key->ipv4.daddr = skops->local_ip4;
-        key->ipv4.dport = bpf_htonl(skops->local_port) >> 16;
-    }
+	struct kmesh_manager_key key = {0};
+	key.ipaddress = ip;
+	__u8 *value = bpf_map_lookup_elem(&map_of_kmesh_manager, &key);
+	if (value)
+		return 1;
+	return 0;
 }
 
 // clean map_of_auth
 static inline void clean_auth_map(struct bpf_sock_ops *skops)
 {
-    struct bpf_sock_tuple tuple_key = {0};
-    build_auth_key(skops, &tuple_key);
-    long ret = bpf_map_delete_elem(&map_of_auth, &tuple_key);
-    if(ret && ret != -ENOENT) {
-        BPF_LOG(INFO, SOCKOPS, "map_of_auth bpf_map_delete_elem failed, ret: %d\n", ret);
-    }
+	struct bpf_sock_tuple tuple_key = {0};
+	// auth run PASSIVE ESTABLISHED CB now. In thie state cb
+	// tuple info src is server info, dst is client info
+	// During the auth, src must set the client info and dst set
+	// the server info when we transmitted to the kmesh auth info.
+	// In this way, auth can be performed normally.
+	tuple_key.ipv4.saddr = skops->remote_ip4;
+	tuple_key.ipv4.daddr = skops->local_ip4;
+	tuple_key.ipv4.sport = skops->remote_port >> 16;
+	tuple_key.ipv4.dport = bpf_htonl(skops->local_port) >> 16;
+	long ret = bpf_map_delete_elem(&map_of_auth, &tuple_key);
+	if(ret && ret != -ENOENT)
+		BPF_LOG(INFO, SOCKOPS, "map_of_auth bpf_map_delete_elem failed, ret: %d\n", ret);
```

Comment:
log the key to, is there a helper in c to convert struct to string

Reply:
C does not provide the function of converting struct to string by default. We need to implement it. The recorded error codes can be used to locate problems such as out of memory. However, it seems that recording keys is unnecessary because residual keys can be easily found by viewing maps in the system.